src/data/cel.rs

```diff
@@ -0,0 +1,499 @@
+        ("source.labels".into(), ValueType::Map),
+        ("destination.labels".into(), ValueType::Map),
+        ("filter_state".into(), ValueType::Map),
+        ("connection.mtls".into(), ValueType::Bool),
+        ("request.raw_body".into(), ValueType::Bytes),
+    ])
+}
+
+fn properties<'e>(exp: &'e CelExpression, all: &mut Vec<Vec<&'e str>>, path: &mut Vec<&'e str>) {
+    match exp {
+        CelExpression::Arithmetic(e1, _, e2)
+        | CelExpression::Relation(e1, _, e2)
+        | CelExpression::Ternary(e1, _, e2)
+        | CelExpression::Or(e1, e2)
+        | CelExpression::And(e1, e2) => {
+            properties(e1, all, path);
+            properties(e2, all, path);
+        }
+        CelExpression::Unary(_, e) => {
+            properties(e, all, path);
+        }
+        CelExpression::Member(e, a) => {
+            if let Member::Attribute(attr) = &**a {
+                path.insert(0, attr.as_str())
+            }
+            properties(e, all, path);
+        }
+        CelExpression::FunctionCall(_, target, args) => {
+            if let Some(target) = target {
+                properties(target, all, path);
+            }
+            for e in args {
+                properties(e, all, path);
+            }
+        }
+        CelExpression::List(e) => {
+            for e in e {
+                properties(e, all, path);
+            }
+        }
+        CelExpression::Map(v) => {
+            for (e1, e2) in v {
+                properties(e1, all, path);
+                properties(e2, all, path);
+            }
+        }
+        CelExpression::Atom(_) => {}
+        CelExpression::Ident(v) => {
+            if !path.is_empty() {
+                path.insert(0, v.as_str());
+                all.push(path.clone());
+                path.clear();
+            }
+        }
+    }
+}
+
+pub mod data {
+    use crate::data::cel::Attribute;
+    use cel_interpreter::objects::{Key, Map};
+    use cel_interpreter::Value;
+    use std::collections::HashMap;
+    use std::sync::Arc;
+
+    #[derive(Debug)]
+    enum Token {
+        Node(HashMap<String, Token>),
+        Value(Attribute),
+    }
+
+    pub struct AttributeMap {
+        data: HashMap<String, Token>,
+    }
+
+    impl AttributeMap {
+        pub fn new(attributes: Vec<Attribute>) -> Self {
+            let mut root = HashMap::default();
+            for attr in attributes {
+                let mut node = &mut root;
+                let mut it = attr.path.tokens().into_iter();
+                while let Some(token) = it.next() {
+                    if it.len() != 0 {
+                        node = match node
+                            .entry(token.to_string())
+                            .or_insert_with(|| Token::Node(HashMap::default()))
+                        {
+                            Token::Node(node) => node,
+                            Token::Value(_) => unreachable!(), // that's a bit of a lie!
```

Comment:
Yes, right now that is NOT supported. 
Ideally the 2nd entry would need to be filtered out. So I was thinking that one day we could sort, so that the shortest path are always first, and subsequent ones can "just" be ignored. 
It's not an issue right now, as you can't do anything with `auth` for instance right now really. Not entirely true, but say nothing useful. Also, that is also always better expressed without requiring the whole shorter path to be resolved. One exception is `request.headers` but that needs special attention anyways.

Reply:
Have you considered building another type of tree structure?

```rust
struct Node {
     value: Option<Attribute>,  
     children: Vec<Node>,
} 

struct Tree {
     roots: Vec<Node>,
}
```

Reply:
To what purpose? I assume you're talking about this input:

```
[
   Attribute(["auth"]),
   Attribute(["auth", "identity")
]
```
yes? If so, let's zoom back for a second. if you access `auth`, I need to hydrate `auth` in full... So that `["auth", "identity"]` is of no use anymore, at the `identity` member was hydrated as part of `auth`, right? 
Or am I missing the use case you are trying to address here?

Reply:
I've ordered and skip an "more precise" attribute when a Value was resolved earlier in the graph.
So if `auth` was resolved, no need to resolve `auth.identity`, as its a member (hopefully) of `auth` that's now fully hydrated, see [here](https://github.com/Kuadrant/wasm-shim/pull/113/commits/dbdf7e10a7f3f9286143d3ad4ad655255f70e9fc)

Reply:
I was thinking on 

```
contains(request.headers, "X-MY-HEADER") && request.headers.OTHER-HEADER == "some-value"
```

But, no need to address this kind of issues now. Nitpicking here.